src/tree/gpu_hist/gradient_based_sampler.cu

```diff
@@ -0,0 +1,225 @@
+      sample = UniformSampling(gpair, dmat);
+      break;
+    default:
+      LOG(FATAL) << "unknown sampling method";
+      sample = {sample_rows_, page_.get(), gpair};
+  }
+  monitor_.StopCuda("Sample");
+  return sample;
+}
+
+// When not sampling, collect all the external memory ELLPACK pages into a single in-memory page.
+void GradientBasedSampler::CollectPages(DMatrix* dmat) {
+  if (page_collected_) {
+    return;
+  }
+
+  size_t offset = 0;
+  for (auto& batch : dmat->GetBatches<EllpackPage>(batch_param_)) {
+    auto page = batch.Impl();
+    size_t num_elements = page_->Copy(batch_param_.gpu_id, page, offset);
+    offset += num_elements;
+  }
+  page_collected_ = true;
+}
+
+GradientBasedSample GradientBasedSampler::NoSampling(common::Span<GradientPair> gpair,
+                                                     DMatrix* dmat) {
+  CollectPages(dmat);
+  return {sample_rows_, page_.get(), gpair};
+}
+
+/*! \brief A functor that calculates the weight of each row as random(0, 1) / abs(grad). */
+struct CalculateWeight : public thrust::binary_function<GradientPair, size_t, float> {
+  const uint32_t seed;
+
+  XGBOOST_DEVICE explicit CalculateWeight(size_t _seed) : seed(_seed) {}
+
+  XGBOOST_DEVICE float operator()(const GradientPair& gpair, size_t i) {
+    if (gpair.GetGrad() == 0) {
+      return FLT_MAX;
+    }
+    thrust::default_random_engine rng(seed);
+    thrust::uniform_real_distribution<float> dist;
+    rng.discard(i);
+    return dist(rng) / fabsf(gpair.GetGrad());
+  }
+};
+
+/*! \brief A functor that returns true if the gradient pair is non-zero. */
+struct IsNonZero : public thrust::unary_function<GradientPair, bool> {
+  XGBOOST_DEVICE bool operator()(const GradientPair& gpair) const {
+    return gpair.GetGrad() != 0 || gpair.GetHess() != 0;
+  }
+};
+
+/*! \brief A functor that clears the row indexes with empty gradient. */
+struct ClearEmptyRows : public thrust::binary_function<GradientPair, size_t, size_t> {
+  XGBOOST_DEVICE size_t operator()(const GradientPair& gpair, size_t row_index) const {
+    if (gpair.GetGrad() != 0 || gpair.GetHess() != 0) {
+      return row_index;
+    } else {
+      return SIZE_MAX;
+    }
+  }
+};
+
+GradientBasedSample GradientBasedSampler::SequentialPoissonSampling(
+    common::Span<xgboost::GradientPair> gpair, DMatrix* dmat) {
+  // Transform the gradient to weight = random(0, 1) / abs(grad).
+  thrust::transform(dh::tbegin(gpair), dh::tend(gpair),
+                    thrust::counting_iterator<size_t>(0),
+                    dh::tbegin(row_weight_),
+                    CalculateWeight(common::GlobalRandom()()));
+  return WeightedSampling(gpair, dmat);
+}
+
+// Perform sampling after the weights are calculated.
+GradientBasedSample GradientBasedSampler::WeightedSampling(
+      common::Span<xgboost::GradientPair> gpair, DMatrix* dmat) {
+  // Sort the gradient pairs and row indexes by weight.
+  thrust::sort_by_key(dh::tbegin(row_weight_), dh::tend(row_weight_),
```

Comment:
to sample a requisite number of rows from the dataset, i see a number of items that are still pegged to `O(num_rows)` *here*.  for instance, this specific line (radix sort) isn't sorting it in-place - meaning, this is going to need auxiliary device memory to sort the row weights and to associate them with the corresponding gradient pairs and its row indices. this will still perhaps need *at least* O(num_rows) * sizeof(item) additional device memory for each entity being sorted and associated.

a simple back-of-the-envelope calculation for 200m rows will require ~ 3.7gb to hold the row weights, indices etc. ~ 3gb of temp buffer during sort. this will only go much higher for bigger datasets.

perhaps, heap or other sorting algorithm may help, if we can trade time for space. but, even more fundamentally, maintaining the entire meta-data as a function of the size of the dataset in the device memory may not scale for larger ones. perhaps, doing this on host on bringing only the sample row indices may alleviate this, at the expense of speed, if it's tolerable.

Reply:
Yes it's not completely "out of core", but I think we are going to run out of other resources (e.g. main memory) before hitting the GPU memory limit. As it stands now, I actually can't generate a large enough dataset on my desktop to test the external memory mode, had to spin up a high-memory instance on GCP.

Reply:
system memory usage is typically less when external memory mode is enabled, as the entire dataset isn't made resident (specifically during sketching), and only parts are streamed in incrementally. i think it may be worthwhile to check how much more data we can accommodate over and above the gpu memory with this pr. i still think that maintaining the meta data for the dataset in the device memory this way to sample will turn out to be more expensive.

perhaps, the same dataset could be concatenated a few times to create a large dataset just for the purposes of determining the limits.